cpu/sam0_common/periph/rtt.c

```diff
@@ -136,6 +142,7 @@ void rtt_clear_overflow_cb(void)
 uint32_t rtt_get_counter(void)
 {
     _wait_syncbusy();
+    _rtt_read_req();
```

Comment:
does this need to sync, req, sync? isn't sync, read sufficient?

Reply:
It depends, if you assume that all `READREQUEST` operation and write-sync operations are not interrupted then you don't need the first read. But if lets say there was a thread that called `rtt_get_counter()` and then for some reason an interrupt did it at well, then the second read request would stall the bus since the `SYNCBUSY` bit would be set.

Reply:
Honestly the overhead of checking the bit twice to be sure is negligible compared to the 180us the sync operation takes anyway. I'd rather ere on the safe side.

Reply:
> But if lets say there was a thread that called `rtt_get_counter()` and then for some reason an interrupt did it at well, then the second read request would stall the bus since the `SYNCBUSY` bit would be set.

I see, but isn't that still the case? after any 'wait on syncbusy', an isr could interrupt (not unlikely given it takes 180us).

Reply:
is there a register that states if count can be read without delay?

Reply:
Nope, section 17.6.9 specifies that:

> The following registers need synchronization when read:
> z The Counter Value register (COUNT)
> z The Clock Value register (CLOCK)

Reply:
Continous read could be used, but I have tested it and although it has no delay it will only change the value every 180us, so if reading in a loop you would get: `0, 0, 0, 0, 0, 0, 6, 6, 6, 6, 6, 12, 12, 12, 12`. So I think its a worse option.

Reply:
> I see, but isn't that still the case? after any 'wait on syncbusy', an isr could interrupt (not unlikely given it takes 180us).

Yes but interrupts can still them-self get interrupted, depending on the priority, or it could be a ISR that would cause a context switch to another thread that would perform this read, right?

Reply:
so there'll still be a race with '(sync, req), sync, read', if an ISR interrupts and also reads between 'sync, read', that blocks the bus if it hits, right?

Reply:
It would first `wait_for_sync` and since the first caller had set the bit with the `READREQUEST` then that wouldn't happen. Also after a `READREQUEST` is issued and synched, the value stays put until the next `READREQUEST`, so there would be no issue. What might happen is that the first caller might end up reading the synchronized value from the second caller. But that is something the first caller should care for by disabling interrupts when reading the counter and doing sensitive stuff with that value, right?. 

```
      thread1                  thread 2

  rtt_get_counter()               .
    wait_for_sync()               .
    read_request()                .
    wait_for_sync()        rtt_get_counter()   <--------- ctxt switch
         .                    wait_for_sync()
         .                        .
         .                        .
         .                        .
         .                    read_request()
         .                    wait_for_sync()
         .                    returns count
         .                        .
    returns count
```

And again reading a register twice, takes nanoseconds vs 180us that the sync operation or a potential stall would take.

Reply:
```
         .                    returns count
         .                        .
    returns count
```

but, anything might happen after thread 2 gets scheduled. also, "returns count" is actually "reads count". so let's assume thread 2 does something that takes time or make the sync invalid, the "read count" of thread 1 could still stall the bus, right?

Reply:
> Also after a `READREQUEST` is issued and synched, the value stays put until the next `READREQUEST`, so there would be no issue.

ah, this I didn't get. so there's no issue. thanks!